architecture/crds.md

```diff
@@ -0,0 +1,324 @@
+
+Our current solution (Helm 3) supports installing CRDs, but does not support modifying or deleting CRDs. Those two operations currently must be done out of band. The sections below explain this decision.
+
+### Installing CRDs
+
+There are a number of well-described issues with installing CRDs. Users must have broad permissions. A CRD must be fully installed before a manifest that references that CRD can be accepted by the Kubernetes API server. (However, a CR can be accepted before there is anything available to handle the CR event.) It is entirely possible for two different applications to require different versions of the same CRD, but no clear way to support or declare that need within Kubernetes. This is exacerbated by the fact that while a CRD is a global object, the controllers that listen for CRD events may be namespace-bound. This means that two instances of the same controller (in different namespaces) can use the same CRD definition. There is no way to query Kubernetes to discover this fact.
+
+All of these things present challenges for installing CRDs. Essentially, any installation process must have some kind of logic like "If the CRD exists, check to make sure that it is compatible with the version of the CRD that I need, and then proceed on my installation. Otherwise, attempt to install the CRD, and handle the case that I might not have permission to do so."
+
+After looking at several ways of making it easy for chart developers to do this, we alighted on a fairly easy (though inelegant) solution: Helm, as it stands today, provides a separate location in a chart to add CRDs. These will be uploaded _before the chart's templates are even rendered_ (thus avoiding verification issues with the discovery API). If the CRD exists or if the permissions do not allow it, the chart will fail before the application is deployed.
+
+Of course, users are unhappy with this for a host of reasons. They want CRDs templated (without understanding the race conditions). They want stronger version controls. They don't like having a separate directory for CRDs. We sympathize... but we currently have not devised a better solution for the aforementioned problems.
+
```

Comment:
I'll add there is an additional problem. helm3 doesn't prevent crds stored in the template directory. So chart developers either can ignore the crds directory by mistake or purposefully template anyway and cause broken --skip-crds functionality surprising users. some of us know to be weary of --skip-crds is not always protection. But most users don't know this.

This can easily be shown by doing a helm create foo, and downloading any random crd into foo/templates and:
```
$ helm template foo --skip-crds | grep Cus
kind: CustomResourceDefinition
```

I'd argue for helm3 actually filtering out all templates CustomResourceDefinition objects from its actions so it leaves them in place if it already previously rendered into the cluster them but stops owning them like it does for the crds directory. So old releases crds get unowned by helm and then never touched again. Then it behaves like crds dir.

currently I have to carefully check all helm generated content for crds for every upgrade. Helm makes it too easy to sneak in a crd onto a cluster where I can't have it managed by helm.

Reply:
Being able to treat CRDs like any other resource is an [intentional design choice](https://github.com/helm/community/pull/138/files#diff-60ca19ad411dd7a26ef9e4597c3cfb1a4f57a78ae845969d62669347cadbde81R108) per this doc.

That said, it might be worth raising an issue to discuss the interaction between `--skip-crds` and "CRDs as regular templates"; I too have been surprised by this in the past, but the couple of charts where I hit this issue also had exposed a value to allow skipping the template CRDs, so I set that value and went on with my day.